src/ucp/core/ucp_worker.c

```diff
@@ -226,10 +226,87 @@ ucp_to_uct_wakeup_events(ucp_wakeup_event_t ucp_events)
     return uct_events;
 }
 
+static void ucp_ep_err_pending_purge(uct_pending_req_t *self, void *arg)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    ucs_status_t status = (uintptr_t)arg;
+
+    ucp_request_complete_send(req, status);
+}
+
+static void
+ucp_worker_iface_error_handler(void *arg, uct_ep_h uct_ep, ucs_status_t status)
+{
+    ucp_worker_h       worker           = (ucp_worker_h)arg;
+    ucp_ep_h           ucp_ep           = NULL;
+    uint64_t           dest_uuid UCS_V_UNUSED;
+    ucp_ep_h           ucp_ep_iter;
+    khiter_t           ucp_ep_errh_iter;
+    ucp_err_handler_t  err_handler;
+    ucp_rsc_index_t    lane, n_lanes, failed_lane;
+
+    /* TODO: need to optimize uct_ep -> ucp_ep lookup */
+    kh_foreach(&worker->ep_hash, dest_uuid, ucp_ep_iter, {
+        for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep_iter); ++lane) {
+            if (uct_ep == ucp_ep_iter->uct_eps[lane]) {
+                ucp_ep = ucp_ep_iter;
+                failed_lane = lane;
+                goto found_ucp_ep;
+            }
+        }
+    });
+
+    ucs_fatal("No uct_ep_h %p associated with ucp_ep_h on ucp_worker_h %p",
+              uct_ep, worker);
+    return;
+
+found_ucp_ep:
+
+    /* Purge outstanding */
+    uct_ep_pending_purge(uct_ep, ucp_ep_err_pending_purge, (void*)status);
+
+    /* Destroy all lanes excepting failed one since ucp_ep becomes unusable as well */
+    for (lane = 0, n_lanes = ucp_ep_num_lanes(ucp_ep); lane < n_lanes; ++lane) {
+        if ((lane != failed_lane) && ucp_ep->uct_eps[lane]) {
+            /* TODO flush and purge outstanding operations */
+            uct_ep_destroy(ucp_ep->uct_eps[lane]);
+            ucp_ep->uct_eps[lane] = NULL;
+        }
+    }
+
+    /* Move failed lane at 0 index */
+    if (failed_lane != 0) {
+        ucp_ep->uct_eps[0] = ucp_ep->uct_eps[failed_lane];
+        ucp_ep->uct_eps[failed_lane] = NULL;
+    }
+
+    /* Redirect all lanes to failed one */
+    ucp_ep_config_key_t key = ucp_ep_config(ucp_ep)->key;
+    key.am_lane            = 0;
+    key.wireup_lane        = 0;
+    key.rndv_lane          = 0;
+    key.amo_lanes[0]       = 0;
+    key.rma_lanes[0]       = 0;
+    key.lanes[0].rsc_index = UCP_NULL_RESOURCE;
```

Comment:
need to check if operations which do registration (e.g zcopy) fail correctly on a failed ep, and there is no problem with memory registration or cleanup of memory registration